feat: add new notification dispatcher mechanism

linuxbandit opened the pull request:
I tried to account for all the questions in README.md

linuxbandit committed:
feat: add new notification dispatcher mechanism

linuxbandit committed:
feat: add retry if template not found

removing the auto unack, the program will not remove the message
from the queue if it can't process it

linuxbandit committed:
chore: add docker setup

linuxbandit committed:
chore: declare queue only once

linuxbandit committed:
feat: make persistent queue and message

linuxbandit committed:
chore: add explicit exchange as per pattern

linuxbandit committed:
chore: make exchange persistent and add route key

linuxbandit committed:
feat: add full templates tester, add templates

linuxbandit committed:
chore: optimise docker image building; add volumes

linuxbandit committed:
chore: correct core templates

linuxbandit committed:
feat: add SU templates. Also add macros.jinja2

not sure how it was not present at all

linuxbandit committed:
chore: remove comment

linuxbandit committed:
chore: add all templates for SU, working

linuxbandit committed:
chore: add events templates

linuxbandit committed:
chore: add stuff for statutory

WikiRik reviewed (commented):
Agora has been keeping me busy so I didn't have the time yet to play around with it. But this looks great! 

Did you already think about when we want to replace the mailer with this? I guess this might be after we migrated prod server to use the new infrastructure? Since that might also make it easier for us to set up a temporary staging server to test this on a server before it goes to prod.

WikiRik commented:
And additionally just as a note; we're now introducing Python as a language for one of the microservices while the rest all uses JavaScript. I don't think it's a big issue since Python is broadly used and the dispatcher will likely not need to get regular new features, but I did want to mention it. Holds for the inclusion of the terraform scripts as well, and redis in gsuite-wrapper; as we expand and use more technologies it also can make it more difficult for new developers to get properly onboarded. We'll have to keep that balance in mind, but I think we're still on the good side of it. Replacing elixir with python is definitely a better situation for future maintainers

linuxbandit commented on dispatcher/README.md:
```diff
@@ -0,0 +1,46 @@
+In another console, run `python3 helpers/send.py`.
+Control on `appserver.test:8025` the emails sent. It is also possible to control rabbit's stats on `appserver.test:8080`
+
+This method is not dockerised. Currently unsure about adding the docker way.
```
it is dockerised in the end, just the docs need update

linuxbandit commented on dispatcher/dispatcher/main.py:
```diff
@@ -0,0 +1,67 @@
+        # TODO: send a notification to someone about adding a template
+        print(f"template {msg['template']}.jinja2 not found")
+        return
+    # TODO: check if there is an auto-delete after 30 minutes for stuck un-acked messages
```
I think we need to alert someone to add the template. In rabbit we can see the messages that are un-acked (so it was tried to be processed, but then left there)

linuxbandit replied:
i still have to add the mechanism that the messages are rerouted to retry at some point in the future (presumably when "a person" adds the missing template

WikiRik replied:
If it's just the missing template can we not just reject the message and send an error back that the template does not exist?

WikiRik replied:
Same for messages that are missing some values

linuxbandit replied:
ahhh i thought i replied

the point is that for the dispatcher to reject a message it needs to be synchronous (like it is now) whereas the usage of a message queue is exactly to avoid this synchronicity. Which means that as it currently is, the errors are 2: "missing template" (solvable by human intervention, so the message gets requeued with a delay and operator gets informed to add a new template); and "missing parameter" (unsolvable as there's the need to recreate information, so the message is discarded and a new ticket is created for programmers to fix it). Keep in mind, missing parameter means that there was not a thorough integration testing

HassanCehef replied:
I agree with Fabri on the lack of integration testing: I would accept that some emails are never sent and that we quickly add the proper template, thanks to an automated ticket/a message somewhere else. (Not sure what the proper error reporting channel can be)

I don't think that we need a mechanism to resend emails that never got sent in a first version, or ideally ever.

How can we easily develop this integration test? Setup one account and 3 events, and send an application, modify, cancel, confirm, and see that we do get one email for each action?

linuxbandit replied:
hey flooo! 
yes I think it's how you say, bootstrap the 3 containers and send simple API commands after a small provisioning

linuxbandit replied:
also the proper reporting channel would be slack at the beginning i'd say (we have the small wrapper notify.sh iirc)

linuxbandit commented:
> Replacing elixir with python is definitely a better situation for future maintainers

yes that was definitely a consideration. Python is simple and especially suitable because this is a web-worker; not an API. If it needed some HTTP method then I would probably have done it with node like the gsuite-wrapper

About the redis in the wrapper: in the end we don't use it though, right? (which is a bad choice imho, but this is not the thread for discussing)
About terraform: well it's simplifying stuff for us, given that we don't have newcomers (and to be fair, it's less scary than going to the console)

linuxbandit committed:
chore: add last templates

linuxbandit committed:
chore: add checker for thorough testing

linuxbandit committed:
docs: add dockerisation

linuxbandit committed:
chore: forgot another update

WikiRik commented on dispatcher/README.md:
```diff
@@ -0,0 +1,46 @@
+1. [x] include traefik configuration to have the mailhog and rabbit on a subdomain instead of `domain:port`
+1. [x] When RabbitMQ quits or crashes it will forget the queues and messages unless you tell it not to: we need to mark both the queue and messages as durable
+1. [ ] Add auto-retry (DLQ). rabbit is smart and doesn't let me process a message again unless i force it.. https://devcorner.digitalpress.blog/rabbitmq-retries-the-new-full-story/
+1. [ ] add the telegram queue
```
Interact was not the correct word. I just meant receiving messages

linuxbandit committed:
chore: add quickstart

linuxbandit committed:
fix: correct make test

linuxbandit committed:
chore: add extra checks for errors during render

linuxbandit committed:
feat: add requeue with a delay

HassanCehef commented on dispatcher/docker/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+
+COPY ./templates /usr/app/templates
+COPY ./dispatcher/main.py /usr/app/src/main.py
+
+#USER python
+
+#ENV FOO=BAR
```
```suggestion
```
unless it is needed for local dev?

linuxbandit replied:
USER python should be there for sure, as otherwise it's from the root account and it's bad security - I was only cutting corner for development

HassanCehef commented on dispatcher/templates/events_event_updated.jinja2:
```diff
@@ -0,0 +1,81 @@
+                        <td align="center" style="padding: 56px 56px 28px 56px;" valign="top">
+                          <a style="color: #3498DB; text-decoration: none;" href="my.aegee.eu"
+                            target="_blank"><img style="width: 56px; height: 56px; border: 0;" alt="Logo"
+                              src="https://my.aegee.eu/images/logo.png">
```
23.65Kb for an image that we display in a 56*56 pixels box. We should host a more compressed version logo_email.png, or even consider base64'ing the image in the email.

linuxbandit replied:
lulz

HassanCehef commented on dispatcher/helpers/send.py:
```diff
@@ -0,0 +1,246 @@
+creates between 1 and 8 fake emails and puts in the queue
+OR
+tests all templates it finds in the folder
+"""
```
```suggestion

Works by creating a queue that the dispatcher/main.py can read from, without setting up a local RabbitMq. 
"""
```

linuxbandit replied:
I see that you may think that as I define the queues, I am somehow creating them; but this program simply sends to the queues so we still need a local rabbit. I define the queues because it's how rabbit works

HassanCehef replied:
indeed, I thought the rabbit was created in-process in send.py, but no, we expect one to exist already. My suggestion is wrongly worded, but it would still be worth to explain that.

HassanCehef commented on dispatcher/helpers/send.py:
```diff
@@ -0,0 +1,246 @@
+    if(file_count != templates_tested):
+        print()
+        print()
+        print(f"Tested {templates_tested} templates but the folder contains {file_count}")
```
could we add a "os.Exit(1)" here?

I guess what I'm looking for, is a section in the README "how to dev locally?". And I'd say, if you modify the python, then you want to set ALL_TEMPLATES_TEST=false, and RANDOM_AMOUNT_TEST=true, but if you modify the templates, you surely want ALL_TEMPLATES_TEST=true.

When testing all templates, I think we should mention that we catch some errors (invalid jinja syntax, template not found) but that a lot of things are not covered by this test:
* validity of HTML generated: does it look legible? (also on mobile, and without images)
* validity of content: do we send the proper content: does a new candidature email contain the string "                            A new candidature to position"? I wonder if for this purpose (checking the content), some sort of golden tests would be nice: we don't have dynamic (dates) content, while we use random names for the fields, we could make generate_fake_payload static/deterministic.
* finally, that the service calling the dispatcher does pass enough data to the mailer, so that no field is empty.

linuxbandit replied:
i still don't understand your points here, I read them once again but I am lost :) can you rephrase? also your bullet list to me seems unrelated to the first part of the comment

HassanCehef replied:
fair.

>could we add a "os.Exit(1)" here?

right now, send.py always returns 0, regardless of whether tests passed or not. When we detect that one test case failed, we should return to the OS an exit code !=0 to notify it of a test failure. This can then be integrated with other tooling (in a fish terminal, this will show a red prompt for instance, to catch attention)


>I guess what I'm looking for, is a section in the README "how to dev locally?". ...

We should explain the usefulness of each option, and when one wants to use ALL_TEMPLATES_TEST and when one wants to use RANDOM_AMOUNT_TEST.



>When testing all templates,...

I would like to see written what this test can catch, and what this test doesn't cover at all. This way, a developer knows that they need to perform some extra manual testing to ensure that they didn't break anything: this script won't do it all for them (which is totally fine, as long as it's explicit)

HassanCehef commented on dispatcher/templates/events_event_updated.jinja2:
```diff
@@ -0,0 +1,81 @@
+
+<!DOCTYPE html>
```
I just pick this file, but this applies to all: it's a lot of copy and paste, and I guess we'd like to better template this in the future?
It's always the same logic of title/body: 2 to 5 lines change between each file.

The second point, also for a follow-up task, is to remove the massive tableS (plural :sob: ) and use proper HTML to align the content. Our emails are not content heavy, so this makes a nice follow-up item. 

The two points are intertwined, but can be done in independent PRs: the HTML fix ideally would only be done once the HTML code has been factored in a single file.

linuxbandit replied:
keep in mind (but you know, as per your comment in the PR thread) this is only a blind copy paste from the mailer templates that were in elixir

HassanCehef replied:
yup, I did some research after posting this.
I don't expect action here, but I'd leave it open, until tracked in Jira.

HassanCehef reviewed (commented):
comments on testing and on follow-up tasks.
You left TODOs for the error management already: there will go my next batch of questions.

as for improving the templates to make them maintainable, the source is https://github.com/AEGEE/mailer/blob/stable/lib/omsmailer_web/templates/page/confirm_email.html.eex, so I agree with your decision to purely copy them here.

linuxbandit committed:
chore: correct template, add comments

linuxbandit committed:
feat: requeue with dynamic delay

linuxbandit committed:
chore: make USER non-root

linuxbandit commented:
I rebased as I know nobody is working on it (@HassanCehef fyi, just in case you were trying it)

linuxbandit committed:
chore: differentiate between environments

linuxbandit committed:
chore: add everything into a main to be modular

linuxbandit committed:
chore: add loglevels

linuxbandit commented:
with the last commits it's ready to be used now (at least code-wise). I will have to test it with gmail, or anyone can now anyway

linuxbandit committed:
chore: increase clarity of incremental retry block

linuxbandit committed:
feat: add notification system

had to downgrade python of one minor version to make it work
with the simple library slack_notifications

WikiRik commented:
Before I review this; can you also make a PR on core that changes the [mailer function](https://github.com/AEGEE/core/blob/stable/lib/mailer.js) to this? So I can deploy and test it on my own hosted instance. No need to do all the other microservices, they all use the same mailer function so I can just copy-paste it locally

linuxbandit commented:
just pushed the branch `feat/featureflag`, I will give you the tokens you need in private. But I may as well simplify it for you, so look at `feat/featureflag-rik` branch

WikiRik commented on dispatcher/templates/member_expired.jinja2:
```diff
@@ -0,0 +1,71 @@
+
```
Filename should be membership_expired

linuxbandit replied:
I don't find the corresponding entry anywhere in core 🤔

WikiRik replied:
We don't use it in the 'new' core, so I'm fine with removing it. We can add a new template when we add the functionality again

WikiRik reviewed (commented):
I'll go through all mails manually, but I already took a quick look at the file names to see if all mails are present. Just two comments

The new network_board_welcome email is missing; https://github.com/AEGEE/mailer/blob/stable/lib/omsmailer_web/templates/page/network_board_welcome.html.eex (I still need to integrate that into network, but I worked on the email already)

WikiRik commented:
I still need to go through the summeruniversity emails but I did all other submodules. These are the notes of which I'm pretty sure are not this way in current mailer;
- custom.html works for discount codes, but not statutory massmailer. Likely something with markdown HTML, since I think the replacement data is done by statutory (not sure on that)
- events_applied.html has the fields 'AEGEE experience', 'Ideal SU' and 'Motivation' which don't always exist on 'regular' events applications (but all three do for SUs). events_edited.html is fine

Apart from that; in the current mailer empty fields are just empty (as in '') but in dispatcher they show up as 'None'. Likewise 'true' and 'false' are 'True' and 'False'. But that's just python vs javascript. Either way we should make that nicer in a separate PR. It helps that mailhog is now easy to use for development

WikiRik commented:
A bug in summeruniversity also allowed me to test the Slack notification message. I got the following error; `IsADirectoryError: [Errno 21] Is a directory: '/usr/app/src/../tokens.json'`. Might be my fault, might be because Docker volumes can't be a file but have to be a directory. Haven't looked into this

WikiRik commented on dispatcher/templates/candidate_applied.jinja2:
Doing general remarks about the templates in this file, since it's the first one.

my.aegee.eu is hardcoded in a lot of these which we should fix in a future PR so the proper domain is used.

linuxbandit replied:
by proper you mean https, or also a variable (for our dev server)?

WikiRik replied:
Variable

WikiRik commented on dispatcher/templates/candidate_applied.jinja2:
```diff
@@ -0,0 +1,82 @@
+<html lang="en">
+
+{% import "snippets/macros.jinja2" as macros %}
+{{ macros.head("MyAEGEE: New candidature submitted") }}
```
Since these are not being used, I have not reviewed them yet

linuxbandit replied:
i think they are only used by the sender?

WikiRik replied:
I made a related comment, we're using the subjects from constants instead of this title

WikiRik commented on dispatcher/templates/candidate_applied.jinja2:
```diff
@@ -0,0 +1,82 @@
+                    <tbody>
+                      <tr>
+                        <td align="center" style="padding: 28px 56px;" valign="top">
+                          <div style="font-family: &quot;lato&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; line-height: 28px;font-size: 16px; line-height: 24px; color: #A7ADB5;">This email was autogenerated by MyAEGEE mailer.</div>
```
For a future PR; might be nice to reference the helpdesk here for if people have questions

WikiRik commented on dispatcher/templates/confirm_email.jinja2:
```diff
@@ -0,0 +1,83 @@
+                      <tr>
+                        <td align="center" style="padding: 0 56px 28px 56px;" valign="top">
+                          <div style="font-family: &quot;lato&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; line-height: 28px;font-size: 20px; color: #333;"><strong>Hello {{member_firstname}} {{member_surname}} </strong></div>
+                          <div style="font-family: &quot;lato&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; line-height: 28px;font-size: 16px; color: #333;">You've recently registered on AEGEE Intranet. To confirm your account, follow this link</div>
```
```suggestion
                          <div style="font-family: &quot;lato&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; line-height: 28px;font-size: 16px; color: #333;">You've recently registered on MyAEGEE. To confirm your account, follow this link</div>
```

WikiRik commented on dispatcher/templates/custom.jinja2:
```diff
@@ -0,0 +1,2 @@
+<!-- Just takes one parameter and renders that one unescaped -->
+{{ body }}
```
Just to recap. This works fine for code claiming in discounts, but not if I try to send the following in the statutory massmailer;
```md
{first_name} - user's first name
{last_name} - user's last name
{participant_type_order} - applicant's type order, in this format: type (order), or not set if it's not set by the board.
{body_name} - name of the body
{statutory_id} - applicant's statutory id


*italic*
**bold**


# Header 1


## Header 2
```

WikiRik replied:
This was the resulting error in dispatcher;
```
Traceback (most recent call last):
  File "/usr/app/src/main.py", line 275, in <module>
    main()
  File "/usr/app/src/main.py", line 271, in main
    channel.start_consuming()
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/blocking_connection.py", line 1883, in start_consuming
    self._process_data_events(time_limit=None)
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/blocking_connection.py", line 2044, in _process_data_events
    self.connection.process_data_events(time_limit=time_limit)
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/blocking_connection.py", line 851, in process_data_events
    self._dispatch_channel_events()
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/blocking_connection.py", line 567, in _dispatch_channel_events
    impl_channel._get_cookie()._dispatch_events()
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/blocking_connection.py", line 1510, in _dispatch_events
    consumer_info.on_message_callback(self, evt.method,
  File "/usr/app/src/main.py", line 114, in send_email
    rendered = template.render(msg['parameters'], altro=msg['subject'])
  File "/usr/local/lib/python3.9/site-packages/jinja2/environment.py", line 1286, in render
    ctx = self.new_context(dict(*args, **kwargs))
ValueError: dictionary update sequence element #0 has length 1; 2 is required
```

linuxbandit replied:
thanks, I will have a look

WikiRik commented on dispatcher/templates/snippets/events_application_info.jinja2:
```diff
@@ -0,0 +1,12 @@
+  <li><strong>Email: </strong>{{ application["email"] }}</li>
+
+  {% for question in event["questions"] %}
+    <li><strong>{{ question["description"] }} </strong> {{ application["answers"][loop.index0] }}</li>
```
```suggestion
    <li><strong>{{ question["description"] }}: </strong> {{ application["answers"][loop.index0] }}</li>
```

WikiRik commented on dispatcher/templates/snippets/statutory_application_info.jinja2:
```diff
@@ -0,0 +1,30 @@
+  <li><strong>Embassy: </strong>{{ application["visa_embassy"] }}</li>
+
+  {% for question in event["questions"] %}
+    <li><strong>{{ question["description"] }} </strong> {{ application["answers"][loop.index0] }}</li>
```
```suggestion
    <li><strong>{{ question["description"] }}: </strong> {{ application["answers"][loop.index0] }}</li>
```

WikiRik commented on dispatcher/templates/snippets/summeruniversity_application_info.jinja2:
```diff
@@ -0,0 +1,15 @@
+  <li><strong>Motivation: </strong>{{ application["motivation"] }}</li>
+
+  {% for question in event["questions"] %}
+    <li><strong>{{ question["description"] }} </strong> {{ application["answers"][loop.index0] }}</li>
```
```suggestion
    <li><strong>{{ question["description"] }}: </strong> {{ application["answers"][loop.index0] }}</li>
```

WikiRik commented on dispatcher/templates/snippets/statutory_application_info.jinja2:
```diff
@@ -0,0 +1,30 @@
+  <li><strong>Allergies: </strong>{{ application["allergies"] }}</li>
+  <li><strong>Number of events visited: </strong>{{ application["number_of_events_visited"] }}</li>
+  <li><strong>Visa required?: </strong>{{ application["visa_required"] }}</li>
+  <li><strong>Visa passport number: </strong>{{ application["visa_passport_number"] }}</li>
+  <li><strong>Visa passport issue date: </strong>{{ application["visa_passport_issue_date"] }}</li>
+  <li><strong>Visa passport expiration date: </strong>{{ application["visa_passport_expiration_date"] }}</li>
+  <li><strong>Visa passport issue authority: </strong>{{ application["visa_passport_issue_authority"] }}</li>
+  <li><strong>Place of birth: </strong>{{ application["visa_place_of_birth"] }}</li>
+  <li><strong>Street and house number: </strong>{{ application["visa_street_and_house"] }}</li>
+  <li><strong>Postal code: </strong>{{ application["visa_postal_code"] }}</li>
+  <li><strong>City: </strong>{{ application["visa_city"] }}</li>
+  <li><strong>Country: </strong>{{ application["visa_country"] }}</li>
+  <li><strong>Embassy: </strong>{{ application["visa_embassy"] }}</li>
```
For a future PR; we should only show this if visa is required.

WikiRik commented on dispatcher/templates/snippets/board_info.jinja2:
```diff
@@ -0,0 +1,12 @@
+  <li><strong>End date: </strong>{{ board["end_date"] }}</li>
+
+  {% for position in positions %}
+    <li><strong>{{ position["function"] }} </strong> {{ position["name"] }}</li>
```
```suggestion
    <li><strong>{{ position["function"] }}: </strong> {{ position["name"] }}</li>
```

WikiRik commented on dispatcher/templates/member_joined.jinja2:
```diff
@@ -0,0 +1,77 @@
+                      <tr>
+                        <td align="center" style="padding: 0 56px 28px 56px;" valign="top">
+                          <div style="font-family: &quot;lato&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; line-height: 28px;font-size: 20px; color: #333;"><strong>You might have a new member soon!</strong></div>
+                          <div style="font-family: &quot;lato&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; line-height: 28px;font-size: 16px; color: #333;">Recently member {{member_firstname}} {{member_surname}} requested to join your body {{body_name}} </div>
```
```suggestion
                          <div style="font-family: &quot;lato&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; line-height: 28px;font-size: 16px; color: #333;">{{member_firstname}} {{member_surname}} requested to join {{body_name}} </div>
```

WikiRik commented on dispatcher/templates/events_submitted.jinja2:
```diff
@@ -0,0 +1,81 @@
+                          <div style="font-family: &quot;lato&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; line-height: 28px;font-size: 20px; color: #333;"><strong>A new event is submitted</strong></div>
+                          <div style="font-family: &quot;lato&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; line-height: 28px;font-size: 16px; color: #333;">
+                            A new event
+                            <strong>"{{ event["name"] }}"</strong>,
```
```suggestion
                            <strong>"{{ event["name"] }}"</strong>
```

WikiRik commented on dispatcher/templates/events_status_changed.jinja2:
```diff
@@ -0,0 +1,82 @@
+                        <td align="center" style="padding: 0 56px 28px 56px;" valign="top">
+                          <div style="font-family: &quot;lato&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; line-height: 28px;font-size: 20px; color: #333;"><strong>Your event's status was changed</strong></div>
+                          <div style="font-family: &quot;lato&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; line-height: 28px;font-size: 16px; color: #333;">
+                            The status of your event,
+                            <strong>"{{ event["name"] }}"</strong>,
```
```suggestion
                            The status of your event,
                            <strong>{{ event["name"] }}</strong>,
```

We do this in other instances as well, they can all be changed

WikiRik commented on dispatcher/templates/events_status_changed.jinja2:
```diff
@@ -0,0 +1,82 @@
+                            The status of your event,
+                            <strong>"{{ event["name"] }}"</strong>,
+                            has changed from <strong>{{ old_status }}</strong>
+                            to <strong> {{ event["status"] }}</strong>.
```
```suggestion
                            to <strong>{{ event["status"] }}</strong>.
```

WikiRik commented on dispatcher/templates/snippets/events_application_info.jinja2:
```diff
@@ -0,0 +1,12 @@
+<ul>
+  <li><strong>Body name: </strong>{{ application["body_name"] }}</li>
+  <li><strong>Applied on: </strong>{{ application["created_at"] }}</li>
+  <li><strong>Updated on: </strong>{{ application["updated_at"] }}</li>
```
Future PR; would be nice to format the datetime string

WikiRik commented on dispatcher/templates/events_applied.jinja2:
```diff
@@ -0,0 +1,76 @@
+                        <td style="padding: 0 56px 28px 56px;" valign="top">
+                          <div style="font-family: &quot;lato&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; line-height: 28px;font-size: 16px; color: #333;">
+                            Here are your application details:
+                            {% include "snippets/summeruniversity_application_info.jinja2" %}
```
```suggestion
                            {% include "snippets/events_application_info.jinja2" %}
```

WikiRik commented on dispatcher/templates/statutory_memberslist_edited.jinja2:
```diff
@@ -0,0 +1,71 @@
+                      <tr>
+                        <td align="center" style="padding: 0 56px 28px 56px;" valign="top">
+                          <div style="font-family: &quot;lato&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; line-height: 28px;font-size: 16px; color: #333;">
+                            <strong>"{{ body_name }}"</strong> has edited their memberslist for "{{ event_name }}".
+                            Their new membership fee is "{{ membership_fee }}".
```
```suggestion
                            <strong>{{ body_name }}</strong> has edited their memberslist for {{ event_name }}.
                            Their new membership fee is €{{ membership_fee }}.
```

linuxbandit replied:
not everyone uses euros

linuxbandit replied:
i don't think we normalise in euros considering the exchange rates are another potential headache for a-eu

WikiRik replied:
This is for the fee towards AEGEE-Europe which is normalized in euros

WikiRik commented on dispatcher/templates/statutory_memberslist_submitted.jinja2:
```diff
@@ -0,0 +1,71 @@
+                      <tr>
+                        <td align="center" style="padding: 0 56px 28px 56px;" valign="top">
+                          <div style="font-family: &quot;lato&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; line-height: 28px;font-size: 16px; color: #333;">
+                            <strong>"{{ body_name }}"</strong> has submitted their memberslist for "{{ event_name }}".
+                            Their membership fee is "{{ membership_fee }}".
```
```suggestion
                            <strong>{{ body_name }}</strong> has submitted their memberslist for {{ event_name }}.
                            Their membership fee is €{{ membership_fee }}.
```

WikiRik commented on dispatcher/templates/summeruniversity_application_status_updated.jinja2:
```diff
@@ -0,0 +1,78 @@
+                        <td style="padding: 0 56px 28px 56px;" valign="top">
+                          <div style="font-family: &quot;lato&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; line-height: 28px;font-size: 16px; color: #333;">
+                            The status of your application for {{event["name"]}} is <strong>{{application["status"]}}</strong>.
+                            If you have been accepted, the organizers of the Summer University will be in touch with you soon and provide you with more information.
+                            If you have been rejected, there might be other Summer Universities with open calls. If there are any at this moment you can find them on MyAEGEE.
```
For future PR; we should make this conditional

WikiRik commented on dispatcher/templates/summeruniversity_board_applied.jinja2:
```diff
@@ -0,0 +1,84 @@
+                        <td align="center" style="padding: 0 56px 28px 56px;" valign="top">
+                          <div>
+                            <a href="https://my.aegee.eu/summeruniversity/boardview" style="background-color:#1468C5;border-radius:50px;color:#ffffff;display:inline-block;font-family: 'lato', 'Helvetica Neue', Helvetica, Arial, sans-serif;font-size:18px;font-weight: bold;line-height:40px;text-align:center;text-decoration:none;width:270px;-webkit-text-size-adjust:none;"
+                              target="_blank">Go to the Summer University board view</a>
```
Future PR; the length of this text is not good for the button

WikiRik commented on dispatcher/templates/summeruniversity_status_changed.jinja2:
```diff
@@ -0,0 +1,82 @@
+                          <div style="font-family: &quot;lato&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; line-height: 28px;font-size: 20px; color: #333;"><strong>Your Summer University's status was changed</strong></div>
+                          <div style="font-family: &quot;lato&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; line-height: 28px;font-size: 16px; color: #333;">
+                            The status of your Summer University,
+                            <strong>"{{event["name"]}}"</strong>,
```
```suggestion
                            <strong>{{event["name"]}}</strong>,
```

WikiRik commented on dispatcher/templates/summeruniversity_submitted.jinja2:
```diff
@@ -0,0 +1,87 @@
+                          <div style="font-family: &quot;lato&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; line-height: 28px;font-size: 20px; color: #333;"><strong>A new event is submitted for {{event["status"]}}</strong></div>
+                          <div style="font-family: &quot;lato&quot;, &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; line-height: 28px;font-size: 16px; color: #333;">
+                            A new event
+                            <strong>"{{event["name"]}}"</strong>,
```
```suggestion
                            <strong>{{event["name"]}}</strong>
```

WikiRik commented on dispatcher/tests/__init__.py:
Empty file, was this commited on purpose?

linuxbandit replied:
yes it's python stuff/requirements

WikiRik reviewed (commented):
Comments about the various templates. Most comments are issues in current mailer as well or things for future PRs. I'll try to make some Jira issues from those later

linuxbandit commented on dispatcher/helpers/send.py:
```diff
@@ -0,0 +1,276 @@
+    return f"Agora {faker.sentence(nb_words=1)}"
+
+# from constants.js (at least for core..)
+MAIL_SUBJECTS = {
```
good catch! They are misaligned. But also, keep in mind that the macros.head gives the HTML title of the email, which of course should be the subject. However if they're not in sync, it is not a problem; also as a reminder (and this is valid for some other comments below), I was merely copy-pasting-translating the already-inefficient templates

linuxbandit commented on dispatcher/templates/candidate_applied.jinja2:
```diff
@@ -0,0 +1,82 @@
+                    <tbody>
+                      <tr>
+                        <td align="center" style="padding: 56px 56px 28px 56px;" valign="top">
+                          <a style="color: #3498DB; text-decoration: none;" href="my.aegee.eu"
```
i left it out of the batch because it's a bigger change

linuxbandit committed:
chore: apply suggestions from code review for templates

Co-authored-by: Rik Smale <13023439+WikiRik@users.noreply.github.com>

linuxbandit commented:
The discussion is getting huge :D 

in short what i have to do is to check 
[this](https://github.com/AEGEE/MyAEGEE/pull/975#issuecomment-1741703808) and [this](https://github.com/AEGEE/MyAEGEE/pull/975#issuecomment-1741308111) right? because for the rest, it's optimisation of the templates that can wait another PR

Do i forget anything?

WikiRik commented:
I believe so, the massmailer and token.json file/folder. And then in the python files there's cleanup and error handling and other TODOs still there. Not sure how much you want to tackle in this PR

linuxbandit committed:
feat: improve mgmt of test credentials

linuxbandit committed:
feat: improve testing and docs via makefile

linuxbandit committed:
chore: bump versions to non-EOL

also solves CVE

linuxbandit committed:
chore: lint and clarify

linuxbandit committed:
docs: update README

linuxbandit committed:
feat: add rate limiting for protecting apprise

linuxbandit committed:
feat: validate message before re-enqueuing

linuxbandit committed:
docs: add explanation of last 2 commits

linuxbandit committed:
feat: add docker healthchecks

linuxbandit committed:
feat: improve code robustness

linuxbandit committed:
chore: add explicit credentials

linuxbandit committed:
feat: finish parity with mailer

linuxbandit committed:
chore: avoid 'None' if empty field

linuxbandit commented:
done everything but the tokens - we don't need them anymore anyway

linuxbandit committed:
chore(ci): add linter

linuxbandit commented:
ah forgot to say - this has feature parity with mailer

linuxbandit committed:
feat(queue): upgrade rabbitmq

linuxbandit committed:
chore(alert): rename alerting function

linuxbandit committed:
lint: run ruff

linuxbandit committed:
Merge branch 'stable' into feat/add-notification-dispatcher

WikiRik commented on .env.example:
```diff
@@ -40,6 +40,11 @@ SENDGRID_SMTP_HOST=smtp.sendgrid.net
 SENDGRID_USER=
 SENDGRID_PW=5ecr3t
 
```
`RABBITMQ_USER` and `RABBITMQ_PASS` are missing right now, but we should set those on prod

Commit:
fix(dispatcher): sync templates with mailer

Commit:
chore(dispatcher): add dispatcher to commitlint config

WikiRik committed:
Apply suggestions from code review

Co-authored-by: Rik Smale <13023439+WikiRik@users.noreply.github.com>

linuxbandit committed:
docs(env.example): add user/pw for rabbit and fix subdomain var name

linuxbandit committed:
fix: naming conventions

Commit:
fix(dispatcher): add autoescape and strictundefined

Commit:
ci(dispatcher): add test to CI

WikiRik committed:
Merge branch 'stable' into feat/add-notification-dispatcher
